core/domain/app_feedback_report_validators_test.py

```diff
@@ -0,0 +1,728 @@
+
+    USER_1_EMAIL = 'some@email.com'
+    USER_1_USERNAME = 'username1'
+    # The timestamp in sec since epoch for Mar 7 2021 21:17:16 UTC.
+    REPORT_SUBMITTED_TIMESTAMP = datetime.datetime.fromtimestamp(1615151836)
+    # The timestamp in sec since epoch for Mar 19 2021 17:10:36 UTC.
+    TICKET_CREATION_TIMESTAMP = datetime.datetime.fromtimestamp(1616173836)
+    TICKET_CREATION_TIMESTAMP_MSEC = utils.get_time_in_millisecs(
+        TICKET_CREATION_TIMESTAMP)
+
+    PLATFORM_ANDROID = 'android'
+    PLATFORM_WEB = 'web'
+    TICKET_ID = '%s.%s.%s' % (
+        'random_hash', TICKET_CREATION_TIMESTAMP_MSEC,
+        '16CharString1234')
+    REPORT_TYPE_SUGGESTION = 'suggestion'
+    CATEGORY_OTHER = 'other'
+    PLATFORM_VERSION = '0.1-alpha-abcdef1234'
+    COUNTRY_LOCALE_CODE_INDIA = 'in'
+    ANDROID_DEVICE_MODEL = 'Pixel 4a'
+    ANDROID_SDK_VERSION = 28
+    ENTRY_POINT_NAVIGATION_DRAWER = 'navigation_drawer'
+    TEXT_LANGUAGE_CODE_ENGLISH = 'en'
+    AUDIO_LANGUAGE_CODE_ENGLISH = 'en'
+    ANDROID_REPORT_INFO = {
+        'user_feedback_other_text_input': 'add an admin',
+        'event_logs': ['event1', 'event2'],
+        'logcat_logs': ['logcat1', 'logcat2'],
+        'package_version_code': 1,
+        'language_locale_code': 'en',
+        'entry_point_info': {
+            'entry_point_name': 'crash',
+        },
+        'text_size': 'MEDIUM_TEXT_SIZE',
+        'download_and_update_only_on_wifi': True,
+        'automatically_update_topics': False,
+        'is_admin': False
+    }
+    ANDROID_REPORT_INFO_SCHEMA_VERSION = 1
+
+    def setUp(self):
+        super(AppFeedbackReportModelValidatorTests, self).setUp()
+        self.signup(self.USER_1_EMAIL, self.USER_1_USERNAME)
+        self.user_1_id = self.get_user_id_from_email(self.USER_1_EMAIL)
+        self.report_id = (
+            app_feedback_report_models.AppFeedbackReportModel.create(
+                platform=self.PLATFORM_ANDROID,
+                submitted_on=self.REPORT_SUBMITTED_TIMESTAMP,
+                report_type=self.REPORT_TYPE_SUGGESTION,
+                category=self.CATEGORY_OTHER,
+                platform_version=self.PLATFORM_VERSION,
+                android_device_country_locale_code=(
+                    self.COUNTRY_LOCALE_CODE_INDIA),
+                android_sdk_version=self.ANDROID_SDK_VERSION,
+                android_device_model=self.ANDROID_DEVICE_MODEL,
+                entry_point=self.ENTRY_POINT_NAVIGATION_DRAWER,
+                entry_point_topic_id=None,
+                entry_point_story_id=None,
+                entry_point_exploration_id=None,
+                entry_point_subtopic_id=None,
+                text_language_code=self.TEXT_LANGUAGE_CODE_ENGLISH,
+                audio_language_code=self.AUDIO_LANGUAGE_CODE_ENGLISH,
+                android_report_info=self.ANDROID_REPORT_INFO,
+                web_report_info=None))
+
+        self.expected_output = [
```

Comment:
This probably should be something like ``expected_successful_validation_output`` or similar, since the expected output depends on the check (and sometimes we expect a failure).

Reply:
Ah I don't think we can specify a successful or failure since some of the outputs combine successes with failures (see `test_model_validation_with_invalid_id_fails`), but will change it to 'expected_validation_output' to be more explicit.

Reply:
Isn't it the case that this is the output we expect when there aren't errors? Maybe instead: ``expected_validation_output_no_errors`` would be clearer, then.

Reply:
Maybe this isn't the best way to be explicit with the tests but some cases (like [this one](https://github.com/oppia/oppia/pull/12298/files/47d252ebbb6e14fa93126eb774c81e0b21c08076#diff-4ffb0d53ba231e0dd4db9ac88724116ecfaba81312d8adc6fe94b5c1d7c1a30bR168-R172)) adds new invalid models to storage which makes the expected output a mix of valid and invalid reports so it's not restricted to a "no errors" output. In retrospect I was just mirroring existing validator tests so I'm not sure whether this is necessarily the best way to write these tests.

Reply:
I'm not sure I follow exactly. The reason I'm suggesting the rename is because you're setting a suite-wide constant called "expected_validation_output". Each test has its own expected output, so it's not clear why a test uses its own versus the common expected output. By clarifying in the name _what_ that output it is may result in avoiding this potential area for confusion.

Did I misunderstand something maybe?

Reply:
No, this makes sense. What I mean is that it would also be incorrect to be validating on an `expected_successful_validation_output` when the actual output might not be fully successful.

That said, I see now what you mean and agree that this isn't useful since it's a suite-wide value that represents a single successful validation output. Will update.

Reply:
Ah, I see per the latest changes that I was a bit confused on what you were trying to convey with regards to errors. Please see latest comment, otherwise this thread seems resolved.